add actionsComponent to sections

nickjvm opened the pull request:
When sections are present, wrapping label/actions in `<Header/>` component, based on previous PR feedback.

nickjvm committed:
add actionsComponent to sections

ericsoderberghp commented:
Thanks for responding to the prior comments.

In thinking about this change, I think we should do it in a way that does not pass the items to the Actions from within grommet-index. Consider cases where we have thousands of items that match the section criteria but we've only loaded a subset of them into the UI. Because the actions are associated with the section, they shouldn't be dependent on how much infinite scrolling the user has done to pull items into the UI. Therefore, I think the caller should be passing a specific element for each section instead of a component at the Index level. This element would typically have the section criteria included in its properties to enable it to determine the matching items to perform an action against.

This would allow for bulk actions to be instantiated with the back end by passing the criteria instead of individual instances.

ericsoderberghp commented on src/js/components/Tiles.js:
```diff
@@ -195,14 +207,23 @@ export default class IndexTiles extends Component {
+          size={this.props.size}>
+          {tiles}
+        </Tiles>
+        {Actions && <Actions items={result.items}/>}
```
Instead of putting this to the right of the Tiles. I think we should always have a section header, even when there aren't separate sections defined. Tracy, Chris, and I also discussed moving the two numbers above and below the filter control icon into this section header. That would make it clearer what the bulk actions correspond to.

nickjvm commented:
I understand the reason, but don't quite understand how you're suggesting to implement. Can you provide an example of what you're suggesting?

nickjvm commented:
Thinking about this a little further, this may not be how we want it to work. For example, if the action is "delete," we don't want the user to think they're deleting only the 10 visible items when really they're deleting 100 in the background. Performance is another concern - it'll take much longer to act on a full section of 100 items than it will to act on only the 10 visible.

ericsoderberghp commented:
Here's the idea from the caller's point of view:

```
const CONFIG = {
  ...
  attributes: [
    ...
    {name: 'state', label: 'State', sort: {
      direction: 'asc',
      sections: [
        {label: 'Online', value: 'Online',
          actions: <VirtualMachineSectionActions query="state:online" />},
        {label: 'Offline', value: 'Offline',
          actions: <VirtualMachineSectionActions query="state:online" />},
        {label: 'Unknown'}
      ]
    }},
    ...
  ],
  ...
};
```

ericsoderberghp commented:
Regarding what the expected behavior should be, if the section label is "Offline", then I would expect the actions to apply to the items matching that label. In many situations, not all items will be visible and the label is the more sure reference for what the actions apply to.

Regarding performance, the user should be in control of his experience. If the user wants to perform the action against only 10, he should be able to do that. It's up to the user to decide how to narrow the results down to the 10 he wants. If the user wants to perform it against 100, he should be able to. If necessary, the back-end should be throttling task execution to avoid contention.

nickjvm commented:
How would you anticipate the caller to pass actions when the view isn't separated by sections?

Commit:
letting caller define query for section

ericsoderberghp commented:
Chris, Tracy, and I discussed always having at least one section header. In this header, we'd put the total and filtered counts and remove them from around the filter control. We'd probably label it "All" if there is no search or filter and "Filtered" or "Matching" when a search and/or filter is applied. In this case, I'd expect the caller to add the actions element at the top of the config object, which would only be used if there were no sections in the current sort.

Also, I'm thinking it should really be the back-end or reducers job to break up the items into sections, instead of having the grommet-index Tiles component doing it. So, we should probably change utils/PropTypes result to have a "sections" property:

```
result: PropTypes.shape({
    total: PropTypes.number.isRequired,
    unfilteredTotal: PropTypes.number,
    start: PropTypes.number,
    count: PropTypes.number.isRequired,
    items: PropTypes.arrayOf(PropTypes.object).isRequired,
    error: PropTypes.string,
    sections: PropTypes.shape({
      actions: PropTypes.node,
      count: PropTypes.number.isRequired,
      items: PropTypes.arrayOf(PropTypes.object).isRequired,
      label: PropTypes.node.isRequired,
      start: PropTypes.number,
      total: PropTypes.number.isRequired,
    })
  })
```

nickjvm commented:
What are your thoughts on having the sectioning refactor be a separate pull request? Its really a different scope than the bulk operations and I'd really like to have something in against that task this week.

ericsoderberghp commented:
I think we can add the ability to have an 'actions' property on each section object in the index configuration and also an 'action' property now and do the refactoring of the 'result' property later.

I still think the caller should provide a react element with whatever it needs to know its scope and not a component/function that takes an 'items' property from Index.

Commit:
add actions propType to section shape

syntax cleanup to match existing patterns

syntax cleanup to match existing patterns

nickjvm commented:
Great! This should be ready for review, then.

ericsoderberghp commented on src/js/components/Tiles.js:
```diff
@@ -125,6 +126,8 @@ export default class IndexTiles extends Component {
       }
       let tiles = [];
+      let sectionItems = [];
+      const actions = section.actions;
```
unused?

ericsoderberghp commented on src/js/components/Tiles.js:
```diff
@@ -142,6 +145,7 @@ export default class IndexTiles extends Component {
             selectionIndex = tiles.length;
           }
           tiles.push(this._renderTile(item));
+          sectionItems.push(item);
```
unused?

ericsoderberghp commented on src/js/components/Tiles.js:
```diff
@@ -196,13 +203,18 @@ export default class IndexTiles extends Component {
-        {tiles}
-      </Tiles>
+      <div>
+        {actions && <Header size="small" justify="end" responsive={false} pad={{horizontal: 'small'}}>
```
A very minor item. This seems like a bit too much javascript logic inside the JSX. I think we should move this outside the return().

ericsoderberghp commented on src/scss/grommet-index/_objects.index-tiles.scss:
```diff
@@ -4,9 +4,8 @@
-    padding: halve($inuit-base-spacing-unit) $inuit-base-spacing-unit 0;
-    border-top: 1px solid $border-color;
     color: $secondary-text-color;
+    font-size: $inuit-base-font-size;
```
Should use `@include inuit-font-size($inuit-base-font-size)`.

ericsoderberghp commented:
Thanks for the updates. Very close. Just a few minor things.
